site/content/blog/2019-11-04-unit-testing-mmctl-commands.md

```diff
@@ -0,0 +1,140 @@
+
+The rule of thumb here is to use `assert` when you can still get valuable information from the following assertions of your test, and use `require` if, after an error, the following checks won't make sense. For example, if we get an error opening a file, it doesn't make sense to check the file contents.
+
+This is how the command execution and its assertions look like:
+
+```go
+err := searchUserCmdF(s.client, &cobra.Command{}, []string{emailArg})
+s.Require().Nil(err)
+s.Require().Equal(&mockUser, printer.GetLines()[0])
+s.Require().Len(printer.GetErrorLines(), 0)
+```
+
+In the case of searching for an existing user, the first thing we expect is the error to come back as `nil`, so that's our first thing to assert.
+
+To check the output of the commands, we use the `printer` struct. Every significant message that a command prints goes through this struct, that accumulates the output during the execution. At testing time, we can use these accumulated lines to check that the output matches our expectations.
+
+We have two kinds of output: the lines for stdout and the error lines for stderr. In our case, as the user should have been found and printed through stdout, we can use the `printer.GetLines` method to get a slice with the printed messages, and the `Equal` helper to perform the assertion. Then, as we are expecting no errors, we can check that the `printer` struct printed none during the command execution asserting that `printer.GetErrorLines` is empty.
+
+# Printer cleanup
+
+One thing we need to remember is that the `printer` struct gets cleaned for us between test functions courtesy of `MmctlUnitTestSuite`, but we will need to clean it manually between test cases. Therefore, if we have more than one use case in the same test function, we need to run `printer.Clean()` at the beginning of all the `s.Run` blocks but the first one:
+
+```go
+func (s *MmctlUnitTestSuite) TestMyCommandCmd() {
+    s.Run("First test case", func() {
+        // ...
+    })
+
+    s.Run("First test case", func() {
+        printer.Clean()
+        // ...
+    })
+
+    s.Run("First test case", func() {
+        printer.Clean()
+        // ...
+    })
+}
+```
+
+# Participate on the campaign!
+
+And that's it! That's how you write a unit test for a `mmctl` command. If you are interested on contributing to this Open Source campaign, please check the [Up for Grabs tickets](https://github.com/mattermost/mattermost-server/issues?q=is%3Aissue+is%3Aopen+label%3A%22Up+For+Grabs%22) on our Github repository and search for those referring to `mmctl` (or for any other!). If you see a ticket that you would like to work on, just leave a comment asking for it to be assigned to you.
```

Comment:
What do you think about a `Area/mmtcl` label? That would make searching for them much easier.

Reply:
Love it! Will ping Carlos to create it, thanks for the suggestion!! :)

Reply:
I just created it myself :wink: 

https://github.com/mattermost/mattermost-server/labels/Area%2Fmmtcl

Reply:
Oh! didn't know that you had such powers ;) thanks!!!

Reply:
Added a reference to the `Area/mmctl` label and modified the wrap up to reference tickets non-specific to `mmctl` [in this commit](https://github.com/mattermost/mattermost-developer-documentation/pull/424/commits/564e0631cf12dcf3468e4bef72a9673cba4071a6). @hanzei PTAL :)

Reply:
Well, I think every dev has these powers.